ledger-api/grpc-definitions/com/daml/ledger/api/v1/event_query_service.proto

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+syntax = "proto3";
+
+package com.daml.ledger.api.v1;
+
+import "com/daml/ledger/api/v1/event.proto";
+import "com/daml/ledger/api/v1/value.proto";
+
+option java_outer_classname = "EventQueryServiceOuterClass";
+option java_package = "com.daml.ledger.api.v1";
+option csharp_namespace = "Com.Daml.Ledger.Api.V1";
+
+// Allow clients to query events
+service EventQueryService {
+
+  // Get the create and the consuming exercise event for the contract with the provided ID
+  rpc GetEventsByContractId (GetEventsByContractIdRequest) returns (GetEventsByContractIdResponse);
+
+  // Get all create and consuming exercise events for the contracts with the provided contract key
+  rpc GetEventsByContractKey (GetEventsByContractKeyRequest) returns (GetEventsByContractKeyResponse);
+
+}
+
+message GetEventsByContractIdRequest {
+  // The contract id being queried.
+  // Required
+  string contract_id = 1;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract id will only be returned if the requesting parties includes
+  // at least one party that is a stakeholder of the event. For a definition of stakeholders see
+  // https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#contract-observers-and-stakeholders
+  // Required
+  repeated string requesting_parties = 2;
+
+}
+
+message GetEventsByContractIdResponse {
+
+  // Will be populated if a created event is found for the request.
+  // Optional
+  CreatedEvent create_event = 1;
+
+  // Will be populated if a archive event is found for the `created_event`
+  // Optional
+  ArchivedEvent archive_event = 2;
+
+}
+
+message GetEventsByContractKeyRequest {
+
+  // The contract key to search for.
+  // Required
+  Value contract_key = 1;
+
+  // The template id associated with the contract key
+  // Required
+  Identifier template_id = 2;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract key will only be returned if the requesting parties includes
+  // at least one party that is a stakeholder of the event. For a definition of stakeholders see
+  // https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#contract-observers-and-stakeholders
+  // Required
+  repeated string requesting_parties = 3;
+
+  // If populated the `create_event` in the response will be the last create event to happen prior to this event
+  // If not populated then the create event in the response will be that latest create event.
+  // By working backwards from event to event all historical key events can be discovered.
+  // Optional
+  string end_exclusive_event_id = 4;
```

Comment:
Would you consider using an opaque continuation token instead?
As far as I can tell the usage does not strictly require to use the event id for this purpose, on the other hand it binds the user of the API to use it.
The event id contains the transaction id, which is becoming a dodgy concept in pruned ranges. (TBH I would consider blind that info for returned events, since currently users are not aware of transaction ids before the pruning offset, and this way it could leak out).
Currently this works, since only one continuation can be in the pruned zone, but if that changes, we would be immediately in trouble (for example due to non unique contract keys). Please note the transaction_meta table is pruned too and should not contain entries in the pruned range.
As an opaque continuation token you could use the event id still, or event the event seq id which would simplify the logic more.

Reply:
opaque continuation sounds compelling but has one drawback, you can only work your way back from the very end. You can't jump into the sequence from the middle. This can be problematic for a contract key that is rotated constantly